[dagster-powerbi] Move contextual data from DagsterPowerBITranslator to PowerBITranslatorData

maximearmstrong opened the pull request:
## Summary & Motivation

This PR implements `PowerBITranslatorData`, a container class containing a `PowerBIContentData` object (props) and a `PowerBIWorkspaceData` object (context), to pass contextual data to the translator without breaking the `get_asset_spec` signature or requiring context to be passed to the `__init__` method of the `DagsterPowerBITranslator`.

This is PR is an alternative to the proposals mentioned in https://github.com/dagster-io/dagster/pull/26617.

## How I Tested These Changes

Updated tests with BK

## Changelog

[dagster-powerbi] Type hints in the signature of `DagsterPowerBITranslator.get_asset_spec` have been updated - the parameter `data` is now of type `PowerBITranslatorData` instead of `PowerBIContentData`. Custom Power BI translators should be updated.

maximearmstrong commented:
* **#26654** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26654?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26654?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

schrockn commented:
Is this every going to cause a recursive explosion of data? 

I guess we are relying on the fact that `PowerBIWorkspaceData` will only ever hold references to `PowerBIContentData` rather than `PowerBITranslatorData`.

That's kind of why I was imagining a container class that holds a reference

```python
class PowerBITranslatorData:
   content: PowerBIContentData
   workspace: PowerBIWorkspaceData
```

As it would make the possibility structurally impossible via the type system.

Overall this looks good though.

maximearmstrong commented:
@schrockn in theory, it's not a problem because `PowerBIWorkspaceData` is expected to be created with `PowerBIContentData` objects and not `PowerBITranslatorData` objects.

Currently, `DagsterPowerBITranslator.get_asset_spec` takes a `PowerBIContentData` object - Making `PowerBITranslatorData` a subclass of `PowerBIContentData` avoid breaking the signature.

That said, after thinking about this, we should not subclass `PowerBIContentData` for 2 reasons:
- To make sure we avoid the recursion explosion of data, we would need some type checks in `PowerBIWorkspaceData`, which is kinda awkward
- Adding these type checks is an anti-pattern and breaks the [Liskov substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) in Python - instances of a subclass can be passed where instances of a class are expected without causing any problems.

I updated this PR to make `PowerBITranslatorData` a container class, and updated the signature of `DagsterPowerBITranslator.get_asset_spec` so that the `data` param is of type `PowerBITranslatorData`.

Here's why I think we can move forward with this approach:
- After testing previous custom translators, this change does not raise an exception
- Only type checkers like Pyright will fail with this change, if users are using one and if they use type hints in their custom translator
- The translator classes for BI integrations are not marked as experimental, but every callsite is (spec loader, asset def builder, etc.). Adding a changelog in this PR to state that the type hints in the signature of `DagsterPowerBITranslator.get_asset_spec` changed seems reasonable enough to warn users.

schrockn commented on python_modules/libraries/dagster-powerbi/dagster_powerbi/translator.py:
```diff
@@ -78,6 +78,23 @@ class PowerBIContentData:
     properties: Dict[str, Any]
 
 
+class PowerBITranslatorData(NamedTuple):
```
can just make this `@record` but not serializable to be consistent with the rest of this interface.

maximearmstrong replied:
Done in [7c63440](https://github.com/dagster-io/dagster/pull/26654/commits/7c634402dce3a24b24b92936f1ba763a68870262)

schrockn reviewed (commented):
Container class works for me!

schrockn reviewed (changes requested):
Hey @maximearmstrong I think one thing we could do here to make the transition a little easier would be to implement the properties of `PowerBIContentData` on `PowerBITranslatorData`. That way the code wouldn't break, just the the typehints.

Meaning

```python
class PowerBITranslatorData:
    
   @property
   def content_type(self) -> PowerBIContentType:
      return self.content_data.content_type
```

maximearmstrong commented:
> Hey @maximearmstrong I think one thing we could do here to make the transition a little easier would be to implement the properties of `PowerBIContentData` on `PowerBITranslatorData`. That way the code wouldn't break, just the the typehints.

@schrockn Exactly - that's the current state of this PR. 

The changes requested here were already implemented in [bb5b904](https://github.com/dagster-io/dagster/pull/26654/commits/bb5b9044b3ba8d3c7d90b48e026df9f5f14cd8eb) when we switched to using a container class.

> Only type checkers like Pyright will fail with this change, if users are using one and if they use type hints in their custom translator

I will update this part of this [comment](https://github.com/dagster-io/dagster/pull/26654#issuecomment-2563072053) to emphasize that code does not break in the current state.

dpeng817 commented on examples/project_atproto_dashboard/project_atproto_dashboard/dashboard/definitions.py:
```diff
@@ -21,7 +21,7 @@ class CustomDagsterPowerBITranslator(DagsterPowerBITranslator):
         return (
             super()
-            .get_report_spec(data)
+            .get_report_spec(data)  # type: ignore
```
why new type: ignore?

dpeng817 reviewed (approved):
looks good, let's see if we can get rid of the type ignore in the docs snippet

schrockn reviewed (approved):
Please update PR summary to be accurate (still refers to subclassing)
